client/sniffer/src/test/java/org/elasticsearch/client/sniff/SnifferTests.java

```diff
@@ -0,0 +1,547 @@
+        Sniffer.Task task = sniffer.new Task(randomLongBetween(1, Long.MAX_VALUE));
+
+        ScheduledExecutorService scheduledExecutorService = mock(ScheduledExecutorService.class);
+        ScheduledFuture mockedFuture = mock(ScheduledFuture.class);
+        when(scheduledExecutorService.schedule(any(Runnable.class), any(Long.class), any(TimeUnit.class))).thenReturn(mockedFuture);
+        DefaultScheduler scheduler = new DefaultScheduler(scheduledExecutorService);
+        long delay = randomLongBetween(1, Long.MAX_VALUE);
+        Future<?> future = scheduler.schedule(task, delay);
+        assertSame(mockedFuture, future);
+        verify(scheduledExecutorService).schedule(task, delay, TimeUnit.MILLISECONDS);
+        verifyNoMoreInteractions(scheduledExecutorService, mockedFuture);
+    }
+
+    public void testDefaultSchedulerThreadFactory() {
+        DefaultScheduler defaultScheduler = new DefaultScheduler();
+        try {
+            ScheduledExecutorService executorService = defaultScheduler.executor;
+            assertThat(executorService, instanceOf(ScheduledThreadPoolExecutor.class));
+            assertThat(executorService, instanceOf(ScheduledThreadPoolExecutor.class));
+            ScheduledThreadPoolExecutor executor = (ScheduledThreadPoolExecutor) executorService;
+            assertTrue(executor.getRemoveOnCancelPolicy());
+            assertFalse(executor.getContinueExistingPeriodicTasksAfterShutdownPolicy());
+            assertTrue(executor.getExecuteExistingDelayedTasksAfterShutdownPolicy());
+            assertThat(executor.getThreadFactory(), instanceOf(Sniffer.SnifferThreadFactory.class));
+            int iters = randomIntBetween(3, 10);
+            for (int i = 1; i <= iters; i++) {
+                Thread thread = executor.getThreadFactory().newThread(new Runnable() {
+                    @Override
+                    public void run() {
+
+                    }
+                });
+                assertThat(thread.getName(), equalTo("es_rest_client_sniffer[T#" + i + "]"));
+                assertThat(thread.isDaemon(), is(true));
+            }
+        } finally {
+            defaultScheduler.shutdown();
+        }
+    }
+
+    public void testDefaultSchedulerShutdown() throws Exception {
+        ScheduledThreadPoolExecutor executor = mock(ScheduledThreadPoolExecutor.class);
+        DefaultScheduler defaultScheduler = new DefaultScheduler(executor);
+        defaultScheduler.shutdown();
+        verify(executor).shutdown();
+        verify(executor).awaitTermination(1000, TimeUnit.MILLISECONDS);
+        verify(executor).shutdownNow();
+        verifyNoMoreInteractions(executor);
+
+        when(executor.awaitTermination(1000, TimeUnit.MILLISECONDS)).thenReturn(true);
+        defaultScheduler.shutdown();
+        verify(executor, times(2)).shutdown();
+        verify(executor, times(2)).awaitTermination(1000, TimeUnit.MILLISECONDS);
+        verifyNoMoreInteractions(executor);
+    }
```

Comment:
I wonder whether it makes sense to add a test that uses the sniffer with on failure on against a proper http server. but that would add timing aspects which are not desirable. My concern is that we never test the sniffer in real-life. Maybe it would be nice to turn sniffing on when executing yaml tests?

Reply:
I think we can manage that in a followup.